File: examples/mpi_topology.f90
module mpi_topology

    use mpi

    implicit none

    integer :: mpi_world_cart       !< Communicator for cartesian topology
    integer :: nprocs               !< World size
    integer :: myrank               !< World rank

    !> @brief   Type variable for the information of 1D communicator
    type, public :: cart_comm_1d
        integer :: myrank           !< Rank ID in current communicator
        integer :: nprocs                   !< Number of processes in current communicator
        integer :: west_rank                !< Previous rank ID in current communicator
        integer :: east_rank                !< Next rank ID in current communicator
        integer :: mpi_comm                 !< Current communicator
    end type cart_comm_1d

    type(cart_comm_1d)  :: comm_1d_x1       !< Subcommunicator information in x-direction
    type(cart_comm_1d)  :: comm_1d_x2       !< Subcommunicator information in y-direction
    type(cart_comm_1d)  :: comm_1d_x3       !< Subcommunicator information in z-direction

    public  :: mpi_topology_make
    public  :: mpi_topology_clean

    contains

    !>
    !> @brief       Destroy the communicator for cartesian topology.
    !>
    subroutine mpi_topology_clean()

        implicit none
        integer :: ierr

        call MPI_Comm_free(mpi_world_cart, ierr)

    end subroutine mpi_topology_clean

    !>
    !> @brief       Create the cartesian topology for the MPI processes and subcommunicators.
    !>
   
    subroutine mpi_topology_make()
        use global, only : np1, np2, np3, pbc1, pbc2, pbc3
    
        implicit none
    
        integer :: np_dim(1:3)
        logical :: period(1:3)
        logical :: remain(1:3)
        integer :: ierr
    
        np_dim(1) = np1;    np_dim(2) = np2;    np_dim(3) = np3
        period(1) = pbc1;   period(2) = pbc2;   period(3) = pbc3
    
        call MPI_Cart_create( MPI_COMM_WORLD, 3, np_dim, period, .false., mpi_world_cart, ierr)
    
        remain = [.true., .false., .false.]
        call create_subcommunicator(remain, comm_1d_x1)
    
        remain = [.false., .true., .false.]
        call create_subcommunicator(remain, comm_1d_x2)
    
        remain = [.false., .false., .true.]
        call create_subcommunicator(remain, comm_1d_x3)
    end subroutine mpi_topology_make

    subroutine create_subcommunicator(remain, comm_1d)
        implicit none
        logical, intent(in) :: remain(1:3)
        type(cart_comm_1d), intent(out) :: comm_1d
    
        integer :: ierr
    
        call MPI_Cart_sub( mpi_world_cart, remain, comm_1d%mpi_comm, ierr)
        call MPI_Comm_rank(comm_1d%mpi_comm, comm_1d%myrank, ierr)
        call MPI_Comm_size(comm_1d%mpi_comm, comm_1d%nprocs, ierr)
        call MPI_Cart_shift(comm_1d%mpi_comm, 0, 1, comm_1d%west_rank, comm_1d%east_rank, ierr)
    end subroutine create_subcommunicator

end module mpi_topology

File: 00_C/PaScaL_TDMA/examples/mpi_topology.cpp
//======================================================================================================================
//> @file        mpi_topology.f90
//> @brief       This file contains a module of communication topology for the example problem of PaScaL_TDMA.
//> @details     The target example problem is the three-dimensional time-dependent heat conduction problem 
//>              in a unit cube domain applied with the boundary conditions of vertically constant temperature 
//>              and horizontally periodic boundaries.
//>
//> @author      
//>              - Ki-Ha Kim (k-kiha@yonsei.ac.kr), School of Mathematics and Computing (Computational Science and Engineering), Yonsei University
//>              - Ji-Hoon Kang (jhkang@kisti.re.kr), Korea Institute of Science and Technology Information
//>              - Jung-Il Choi (jic@yonsei.ac.kr), School of Mathematics and Computing (Computational Science and Engineering), Yonsei University
//>
//> @date        May 2023
//> @version     2.0
//> @par         Copyright
//>              Copyright (c) 2019-2023 Ki-Ha Kim and Jung-Il choi, Yonsei University and 
//>              Ji-Hoon Kang, Korea Institute of Science and Technology Information, All rights reserved.
//> @par         License     
//>              This project is release under the terms of the MIT License (see LICENSE file).
//======================================================================================================================

//>
//> @brief       Module for creating the cartesian topology of the MPI processes and subcommunicators.
//> @details     This module has three subcommunicators in each-direction and related subroutines.
//>
#include <mpi.h>
#include "modules.hpp"

namespace mpi_topology {

MPI_Comm mpi_world_cart;
int np_dim[3];
int period[3];

cart_comm_1d comm_1d_x;
cart_comm_1d comm_1d_y;
cart_comm_1d comm_1d_z;

void mpi_topology_clean() {
    MPI_Comm_free(&mpi_world_cart);
}

void mpi_topology_make() {
    MPI_Cart_create(MPI_COMM_WORLD, 3, np_dim, period, 0, &mpi_world_cart);

    int remain[3];

    remain[0] = 1; remain[1] = 0; remain[2] = 0;
    MPI_Cart_sub(mpi_world_cart, remain, &comm_1d_x.mpi_comm);
    MPI_Comm_rank(comm_1d_x.mpi_comm, &comm_1d_x.myrank);
    MPI_Comm_size(comm_1d_x.mpi_comm, &comm_1d_x.nprocs);
    MPI_Cart_shift(comm_1d_x.mpi_comm, 0, 1, &comm_1d_x.west_rank, &comm_1d_x.east_rank);

    remain[0] = 0; remain[1] = 1; remain[2] = 0;
    MPI_Cart_sub(mpi_world_cart, remain, &comm_1d_y.mpi_comm);
    MPI_Comm_rank(comm_1d_y.mpi_comm, &comm_1d_y.myrank);
    MPI_Comm_size(comm_1d_y.mpi_comm, &comm_1d_y.nprocs);
    MPI_Cart_shift(comm_1d_y.mpi_comm, 0, 1, &comm_1d_y.west_rank, &comm_1d_y.east_rank);

    remain[0] = 0; remain[1] = 0; remain[2] = 1;
    MPI_Cart_sub(mpi_world_cart, remain, &comm_1d_z.mpi_comm);
    MPI_Comm_rank(comm_1d_z.mpi_comm, &comm_1d_z.myrank);
    MPI_Comm_size(comm_1d_z.mpi_comm, &comm_1d_z.nprocs);
    MPI_Cart_shift(comm_1d_z.mpi_comm, 0, 1, &comm_1d_z.west_rank, &comm_1d_z.east_rank);
}

} // namespace mpi_topology
